File: frontend/src/lib/geo.ts
// Great-circle geodesy for maritime navigation readouts.
//
// Every navigation value the UI shows (bearing, distance to waypoint,
// progress, ETA) is computed here from the route geometry boreas-core
// returned -- nothing is fabricated or hard-coded.

export const EARTH_RADIUS_KM = 6371.0;
export const KM_PER_NM = 1.852; // exact, by definition

export type LonLat = [number, number];

const toRad = (deg: number) => (deg * Math.PI) / 180;
const toDeg = (rad: number) => (rad * 180) / Math.PI;

/** Great-circle distance in kilometres. */
export function haversineKm(a: LonLat, b: LonLat): number {
  const [lon1, lat1] = a;
  const [lon2, lat2] = b;
  const p1 = toRad(lat1);
  const p2 = toRad(lat2);
  const dPhi = toRad(lat2 - lat1);
  const dLambda = toRad(lon2 - lon1);
  const h =
    Math.sin(dPhi / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dLambda / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(Math.min(1, h)));
}

/** Initial great-circle bearing (true course) in degrees, 0-360. */
export function initialBearingDeg(a: LonLat, b: LonLat): number {
  const [lon1, lat1] = a;
  const [lon2, lat2] = b;
  const p1 = toRad(lat1);
  const p2 = toRad(lat2);
  const dLambda = toRad(lon2 - lon1);
  const y = Math.sin(dLambda) * Math.cos(p2);
  const x = Math.cos(p1) * Math.sin(p2) - Math.sin(p1) * Math.cos(p2) * Math.cos(dLambda);
  return (toDeg(Math.atan2(y, x)) + 360) % 360;
}

/** Point at `fraction` along the great circle from a to b (spherical slerp). */
export function interpolateGreatCircle(a: LonLat, b: LonLat, fraction: number): LonLat {
  const [lon1, lat1] = a;
  const [lon2, lat2] = b;
  const p1 = toRad(lat1);
  const l1 = toRad(lon1);
  const p2 = toRad(lat2);
  const l2 = toRad(lon2);

  const d = haversineKm(a, b) / EARTH_RADIUS_KM;
  if (d < 1e-9) return [lon1, lat1];

  const A = Math.sin((1 - fraction) * d) / Math.sin(d);
  const B = Math.sin(fraction * d) / Math.sin(d);
  const x = A * Math.cos(p1) * Math.cos(l1) + B * Math.cos(p2) * Math.cos(l2);
  const y = A * Math.cos(p1) * Math.sin(l1) + B * Math.cos(p2) * Math.sin(l2);
  const z = A * Math.sin(p1) + B * Math.sin(p2);
  return [toDeg(Math.atan2(y, x)), toDeg(Math.atan2(z, Math.hypot(x, y)))];
}

/** Cumulative distance (km) at each vertex of a path; index 0 is always 0. */
export function cumulativeDistancesKm(path: LonLat[]): number[] {
  const out = [0];
  for (let i = 1; i < path.length; i++) {
    out.push(out[i - 1] + haversineKm(path[i - 1], path[i]));
  }
  return out;
}

export interface PositionOnRoute {
  position: LonLat;
  /** Index of the vertex the vessel has most recently passed. */
  legIndex: number;
  /** Index of the vertex being steered toward. */
  nextWaypointIndex: number;
  distanceTravelledKm: number;
  distanceRemainingKm: number;
  distanceToNextWaypointKm: number;
  /** True course to the next waypoint, degrees. Null once the route is complete. */
  bearingDeg: number | null;
  progressFraction: number;
  isComplete: boolean;
}

/**
 * Resolves a distance-along-route into a geographic position plus the
 * navigation state at that point, by walking the real route vertices.
 */
export function positionAtDistance(
  path: LonLat[],
  cumulative: number[],
  travelledKm: number,
): PositionOnRoute {
  const totalKm = cumulative[cumulative.length - 1];
  const clamped = Math.max(0, Math.min(travelledKm, totalKm));

  if (path.length < 2) {
    return {
      position: path[0] ?? [0, 0],
      legIndex: 0,
      nextWaypointIndex: 0,
      distanceTravelledKm: 0,
      distanceRemainingKm: 0,
      distanceToNextWaypointKm: 0,
      bearingDeg: null,
      progressFraction: 1,
      isComplete: true,
    };
  }

  // Last vertex whose cumulative distance is still behind the vessel.
  let leg = 0;
  while (leg < cumulative.length - 2 && cumulative[leg + 1] <= clamped) leg++;

  const legStartKm = cumulative[leg];
  const legLengthKm = cumulative[leg + 1] - legStartKm;
  const legFraction = legLengthKm > 1e-9 ? (clamped - legStartKm) / legLengthKm : 0;
  const position = interpolateGreatCircle(path[leg], path[leg + 1], legFraction);

  const isComplete = clamped >= totalKm - 1e-6;
  const nextWaypointIndex = Math.min(leg + 1, path.length - 1);

  return {
    position,
    legIndex: leg,
    nextWaypointIndex,
    distanceTravelledKm: clamped,
    distanceRemainingKm: totalKm - clamped,
    distanceToNextWaypointKm: cumulative[nextWaypointIndex] - clamped,
    bearingDeg: isComplete ? null : initialBearingDeg(position, path[nextWaypointIndex]),
    progressFraction: totalKm > 0 ? clamped / totalKm : 1,
    isComplete,
  };
}

// ----------------------------------------------------------------- formatting

/** Maritime coordinate format, e.g. 69° 24.5' S. */
export function formatLatitude(lat: number): string {
  const hemisphere = lat < 0 ? 'S' : 'N';
  const abs = Math.abs(lat);
  const deg = Math.floor(abs);
  const min = (abs - deg) * 60;
  return `${deg}° ${min.toFixed(1).padStart(4, '0')}' ${hemisphere}`;
}

export function formatLongitude(lon: number): string {
  const hemisphere = lon < 0 ? 'W' : 'E';
  const abs = Math.abs(lon);
  const deg = Math.floor(abs);
  const min = (abs - deg) * 60;
  return `${deg}° ${min.toFixed(1).padStart(4, '0')}' ${hemisphere}`;
}

export const kmToNm = (km: number) => km / KM_PER_NM;

/** Compass point for a true bearing, e.g. 214° -> SW. */
export function compassPoint(bearingDeg: number): string {
  const points = ['N', 'NNE', 'NE', 'ENE', 'E', 'ESE', 'SE', 'SSE', 'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW'];
  return points[Math.round(((bearingDeg % 360) / 22.5)) % 16];
}

/** Hours as a maritime duration, e.g. 380.2 -> "15d 20h". */
export function formatDuration(hours: number): string {
  if (!Number.isFinite(hours)) return '—';
  const days = Math.floor(hours / 24);
  const rem = Math.round(hours % 24);
  if (days <= 0) return `${Math.round(hours)}h`;
  return `${days}d ${rem}h`;
}

/** Arrival clock time from now plus `hours`, in UTC. */
export function etaTimestamp(hours: number, from: Date = new Date()): string {
  const arrival = new Date(from.getTime() + hours * 3600_000);
  return `${arrival.toISOString().slice(0, 16).replace('T', ' ')}Z`;
}

File: boreas-core/boreas_core/coordination/geo.py
"""Great-circle geodesy for advancing a vessel's simulated position.

This is a deliberate, function-for-function port of
`frontend/src/lib/geo.ts` (haversine, initial bearing, great-circle slerp,
cumulative distance, position-at-distance): the same geometry the Shore
frontend already uses for its own navigation HUD. Porting it here -- rather
than inventing a different formula -- means the backend's canonical vessel
position (what both Shore and Ship poll) agrees with the frontend's own
math, and it is exactly that: geometry over the real route coordinates
`/mission/plan` returned, not a new prediction model.
"""

import math
from dataclasses import dataclass

EARTH_RADIUS_KM = 6371.0
KM_PER_NM = 1.852  # exact, by definition

LonLat = tuple[float, float]


def haversine_km(a: LonLat, b: LonLat) -> float:
    lon1, lat1 = a
    lon2, lat2 = b
    p1, p2 = math.radians(lat1), math.radians(lat2)
    d_phi = math.radians(lat2 - lat1)
    d_lambda = math.radians(lon2 - lon1)
    h = math.sin(d_phi / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(d_lambda / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(min(1.0, h)))


def initial_bearing_deg(a: LonLat, b: LonLat) -> float:
    lon1, lat1 = a
    lon2, lat2 = b
    p1, p2 = math.radians(lat1), math.radians(lat2)
    d_lambda = math.radians(lon2 - lon1)
    y = math.sin(d_lambda) * math.cos(p2)
    x = math.cos(p1) * math.sin(p2) - math.sin(p1) * math.cos(p2) * math.cos(d_lambda)
    return (math.degrees(math.atan2(y, x)) + 360) % 360


def interpolate_great_circle(a: LonLat, b: LonLat, fraction: float) -> LonLat:
    """Point at `fraction` (0-1) along the great circle from a to b (slerp)."""
    lon1, lat1 = a
    lon2, lat2 = b
    p1, l1 = math.radians(lat1), math.radians(lon1)
    p2, l2 = math.radians(lat2), math.radians(lon2)

    d = haversine_km(a, b) / EARTH_RADIUS_KM
    if d < 1e-9:
        return (lon1, lat1)

    A = math.sin((1 - fraction) * d) / math.sin(d)
    B = math.sin(fraction * d) / math.sin(d)
    x = A * math.cos(p1) * math.cos(l1) + B * math.cos(p2) * math.cos(l2)
    y = A * math.cos(p1) * math.sin(l1) + B * math.cos(p2) * math.sin(l2)
    z = A * math.sin(p1) + B * math.sin(p2)
    return (math.degrees(math.atan2(y, x)), math.degrees(math.atan2(z, math.hypot(x, y))))


def cumulative_distances_km(path: list[LonLat]) -> list[float]:
    out = [0.0]
    for i in range(1, len(path)):
        out.append(out[i - 1] + haversine_km(path[i - 1], path[i]))
    return out


@dataclass(frozen=True)
class PositionOnRoute:
    position: LonLat
    leg_index: int
    next_waypoint_index: int
    distance_travelled_km: float
    distance_remaining_km: float
    distance_to_next_waypoint_km: float
    bearing_deg: float | None
    progress_fraction: float
    is_complete: bool


def position_at_distance(path: list[LonLat], travelled_km: float) -> PositionOnRoute:
    """Resolves a distance-along-route into a geographic position plus
    navigation state, by walking the route's own vertices -- the same
    approach as `frontend/src/lib/geo.ts::positionAtDistance`.
    """
    if len(path) < 2:
        p = path[0] if path else (0.0, 0.0)
        return PositionOnRoute(p, 0, 0, 0.0, 0.0, 0.0, None, 1.0, True)

    cumulative = cumulative_distances_km(path)
    total_km = cumulative[-1]
    clamped = max(0.0, min(travelled_km, total_km))

    leg = 0
    while leg < len(cumulative) - 2 and cumulative[leg + 1] <= clamped:
        leg += 1

    leg_start_km = cumulative[leg]
    leg_length_km = cumulative[leg + 1] - leg_start_km
    leg_fraction = (clamped - leg_start_km) / leg_length_km if leg_length_km > 1e-9 else 0.0
    position = interpolate_great_circle(path[leg], path[leg + 1], leg_fraction)

    is_complete = clamped >= total_km - 1e-6
    next_waypoint_index = min(leg + 1, len(path) - 1)

    return PositionOnRoute(
        position=position,
        leg_index=leg,
        next_waypoint_index=next_waypoint_index,
        distance_travelled_km=clamped,
        distance_remaining_km=total_km - clamped,
        distance_to_next_waypoint_km=cumulative[next_waypoint_index] - clamped,
        bearing_deg=None if is_complete else initial_bearing_deg(position, path[next_waypoint_index]),
        progress_fraction=(clamped / total_km) if total_km > 0 else 1.0,
        is_complete=is_complete,
    )
